airbyte-scheduler/src/main/java/io/airbyte/scheduler/client/SpecCachingSchedulerJobClient.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.scheduler.client;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.airbyte.scheduler.Job;
+import io.airbyte.scheduler.persistence.JobPersistence;
+import java.io.IOException;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This client caches only the results of spec jobs as their output should not change (except in the
+ * case where the docker image is replaced with an image of the same name and tag) and they are
+ * called very frequently.
+ */
+public class SpecCachingSchedulerJobClient extends DefaultSchedulerJobClient implements CachingSchedulerJobClient {
```

Comment:
I usually delegation instead of inheritance for that kind of class:

new SpecCacheSJC(client);

most of the methods would just call:
client.xxx()

and only the one where you want to change behavior you add your code.

That allows you to compose features on top of the DefaultClient

You can argue that this is over-engineered.

Reply:
Also since you can not hide the cache invalidation, I guess it doesn't really matter.